pkg/sqlite/load.go

```diff
@@ -805,3 +805,176 @@ func (s *sqlLoader) addBundleProperties(tx *sql.Tx, bundle *registry.Bundle) err
+			return err
+		}
+	}
+	err = updateChannelEntry.Close()
+	if err != nil {
+		return err
+	}
+
+	stmt, err := tx.Prepare("DELETE FROM channel_entry WHERE operatorbundle_name=?")
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	if _, err := stmt.Exec(csvName); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getTailFromBundle(tx *sql.Tx, name string) (bundles []string, err error) {
+	getReplacesSkips := `SELECT replaces, skips FROM operatorbundle WHERE name=?`
+	isDefaultChannelHead := `SELECT head_operatorbundle_name FROM channel 
+							INNER JOIN package ON channel.name = package.default_channel 
+							WHERE channel.head_operatorbundle_name = ?`
+
+	nameQueue := make(chan string, 100)
+	tail := make(map[string]struct{})
+
+	nameQueue <- name
+
+	for {
+		select {
+		case next := <-nameQueue:
+			rows, err := tx.QueryContext(context.TODO(), getReplacesSkips, next)
+			if err != nil {
+				return nil, err
+			}
+
+			var replaces sql.NullString
+			var skips sql.NullString
+			if rows.Next() {
+				if err := rows.Scan(&replaces, &skips); err != nil {
+					return nil, err
+				}
+			}
+			rows.Close()
+
+			if skips.Valid && skips.String != "" {
+				for _, skip := range strings.Split(skips.String, ",") {
+					tail[skip] = struct{}{}
+				}
+			}
+			if replaces.Valid && replaces.String != "" {
+				// check if replaces is the head of the defaultChannel
+				// if it is, the defaultChannel will be removed
+				// this is not allowed because we cannot know which channel to promote as the new default
+				rows, err := tx.QueryContext(context.TODO(), isDefaultChannelHead, replaces.String)
+				if err != nil {
+					return nil, err
+				}
+				if rows.Next() {
+					var defaultChannelHead sql.NullString
+					err := rows.Scan(&defaultChannelHead)
+					if err != nil {
+						return nil, err
+					}
+					if defaultChannelHead.Valid {
+						return nil, registry.ErrRemovingDefaultChannelDuringDeprecation
+					}
+				}
+				nameQueue <- replaces.String
+				tail[replaces.String] = struct{}{}
+			}
+		default:
+			var allTails []string
+
+			for k := range tail {
+				allTails = append(allTails, k)
+			}
+
+			return allTails, nil
+		}
+	}
+
+}
+
+func getBundleNameAndVersionForImage(tx *sql.Tx, path string) (string, string, error) {
+	query := `SELECT name, version FROM operatorbundle WHERE bundlepath=? LIMIT 1`
+	rows, err := tx.QueryContext(context.TODO(), query, path)
+	if err != nil {
+		return "", "", err
+	}
+	defer rows.Close()
+
+	var name sql.NullString
+	var version sql.NullString
+	if rows.Next() {
+		if err := rows.Scan(&name, &version); err != nil {
+			return "", "", err
+		}
+	}
+	if name.Valid && version.Valid {
+		return name.String, version.String, nil
+	}
+	return "", "", registry.ErrBundleImageNotInDatabase
+}
+
+func (s *sqlLoader) DeprecateBundle(path string) error {
+	tx, err := s.db.Begin()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		tx.Rollback()
```

Comment:
For my betterment, does tx.Commit prevent this call from rollingback the transactions?

Reply:
If tx.Commit is successful it would right? But isn't that what we want?

Reply:
It is what we want - I just wanted to confirm this behavior.